File: app/src/main/java/com/example/mateusz/coffeenotes/BeansTypeListFragment.java
package com.example.mateusz.coffeenotes;

import android.content.Context;
import android.content.Intent;
import android.graphics.Typeface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;
import java.util.UUID;

import static android.app.Activity.RESULT_OK;

public class BeansTypeListFragment extends Fragment {
  private static final String ARG_HIGHLIGHTED_BEANS_TYPE_ID = "highlighted_beans_type_id";
  private static final int EDIT_BEANS_TYPE_REQUEST = 1;

  private RecyclerView beansTypesRecyclerView;
  @Nullable private UUID highlightedBeansTypeId;
  private OnBeansTypeSelectedListener onBeansTypeSelectedListener;

  public BeansTypeListFragment() {
    // Required empty public constructor
  }

  @NonNull
  public static BeansTypeListFragment newInstance(UUID highlightedBeansTypeId) {
    Bundle args = new Bundle();
    args.putSerializable(ARG_HIGHLIGHTED_BEANS_TYPE_ID, highlightedBeansTypeId);
    BeansTypeListFragment beansTypeListFragment = new BeansTypeListFragment();
    beansTypeListFragment.setArguments(args);
    return beansTypeListFragment;
  }

  @NonNull
  public static BeansTypeListFragment newInstance() {
    return new BeansTypeListFragment();
  }

  @Override
  public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    Bundle args = getArguments();
    if (args != null) {
      if (args.containsKey(ARG_HIGHLIGHTED_BEANS_TYPE_ID)) {
        highlightedBeansTypeId = (UUID) args.getSerializable(ARG_HIGHLIGHTED_BEANS_TYPE_ID);
      }
    }
    setHasOptionsMenu(true);
  }

  @Override
  public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                           Bundle savedInstanceState) {
    View view = inflater.inflate(R.layout.fragment_beans_type_list, container, false);

    beansTypesRecyclerView = (RecyclerView) view.findViewById(R.id.beans_types_recycler_view);
    beansTypesRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
    beansTypesRecyclerView.setAdapter(
        new BeansTypeAdapter(BeansTypeDataManager.getInstance().getBeansTypeList()));

    return view;
  }

  @Override
  public void onAttach(Context context) {
    super.onAttach(context);
    try {
      onBeansTypeSelectedListener = (OnBeansTypeSelectedListener) context;
    } catch (ClassCastException e) {
      throw new RuntimeException(
          String.format(
              "%s must implement %s interface",
              context.getClass().getSimpleName(),
              OnBeansTypeSelectedListener.class.getSimpleName()));
    }
  }

  @Override
  public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
    super.onCreateOptionsMenu(menu, inflater);
    inflater.inflate(R.menu.fragment_beans_type_list_menu, menu);
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    switch (item.getItemId()) {
      case R.id.menu_item_new_beans_type:
        Intent intent = BeansTypeActivity.newIntent(getContext());
        startActivityForResult(intent, EDIT_BEANS_TYPE_REQUEST);
        return true;
      default:
        return super.onOptionsItemSelected(item);
    }
  }

  @Override
  public void onActivityResult(int requestCode, int resultCode, Intent data) {
    if (requestCode == EDIT_BEANS_TYPE_REQUEST) {
      if (resultCode == RESULT_OK) {
        beansTypesRecyclerView.getAdapter().notifyDataSetChanged();
      }
    }
  }

  @Override
  public void onResume() {
    super.onResume();
  }

  public interface OnBeansTypeSelectedListener {
    void onBeansTypeSelected(BeansType beansType);
  }

  private class BeansTypeAdapter
      extends RecyclerView.Adapter<BeansTypeAdapter.BeansTypeViewHolder> {
    private List<BeansType> beansTypesList;

    public BeansTypeAdapter(List<BeansType> beansTypesList) {
      this.beansTypesList = beansTypesList;
    }

    public class BeansTypeViewHolder extends RecyclerView.ViewHolder {
      private TextView beansNameTextView;
      private BeansType beansType;

      public BeansTypeViewHolder(@NonNull View itemView) {
        super(itemView);
        beansNameTextView = (TextView) itemView;
        beansNameTextView.setOnClickListener(new View.OnClickListener() {
          @Override
          public void onClick(View v) {
            onBeansTypeSelectedListener.onBeansTypeSelected(beansType);
          }
        });
      }

      public void bindBeansType(BeansType beansType) {
        this.beansType = beansType;
        updateViewHolder();
      }

      private void updateViewHolder() {
        beansNameTextView.setText(beansType.getName());
        if (beansType.getId().equals(highlightedBeansTypeId)) {
          beansNameTextView.setTypeface(null, Typeface.BOLD);
        }
      }
    }

    @NonNull
    @Override
    public BeansTypeAdapter.BeansTypeViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
      LayoutInflater inflater = LayoutInflater.from(getContext());
      View view = inflater.inflate(android.R.layout.simple_list_item_1, parent, false);
      return new BeansTypeViewHolder(view);
    }

    @Override
    public void onBindViewHolder(
        @NonNull BeansTypeAdapter.BeansTypeViewHolder holder, int position) {
      holder.bindBeansType(beansTypesList.get(position));
    }

    @Override
    public int getItemCount() {
      return beansTypesList.size();
    }
  }

}

File: app/src/main/java/com/example/mateusz/coffeenotes/BeansTypeListFragment.kt
package com.example.mateusz.coffeenotes

import android.app.Activity.RESULT_OK
import android.content.Context
import android.content.Intent
import android.graphics.Typeface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.widget.TextView
import com.example.mateusz.coffeenotes.application.MyApplication
import com.example.mateusz.coffeenotes.database.BeansTypeDataManager
import com.example.mateusz.coffeenotes.view.ContentViewHolder
import com.example.mateusz.coffeenotes.view.EditableListFragment
import java.util.*
import javax.inject.Inject

class BeansTypeListFragment : EditableListFragment<BeansType>() {

    @Inject lateinit var beansTypeDataManager: BeansTypeDataManager

    private var highlightedBeansTypeId: UUID? = null

    private lateinit var onBeansTypeSelectedListener: OnBeansTypeSelectedListener

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (activity.application as MyApplication).getAppComponent().inject(this)
        val args = arguments
        if (args != null) {
            if (args.containsKey(ARG_HIGHLIGHTED_BEANS_TYPE_ID)) {
                highlightedBeansTypeId = args.getSerializable(ARG_HIGHLIGHTED_BEANS_TYPE_ID) as UUID
            }
        }
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        try {
            onBeansTypeSelectedListener = context as OnBeansTypeSelectedListener
        } catch (e: ClassCastException) {
            throw RuntimeException(
                    String.format(
                            "%s must implement %s interface",
                            context?.javaClass?.simpleName ?: "[null]",
                            this.javaClass.simpleName))
        }

    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val isConsumed = super.onOptionsItemSelected(item)
        when (item.itemId) {
            R.id.menu_item_list_new_item -> {
                val intent = BeansTypeActivity.newIntent(context)
                startActivityForResult(intent, EDIT_BEANS_TYPE_REQUEST)
                return true
            }
            else -> return isConsumed
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == EDIT_BEANS_TYPE_REQUEST) {
            if (resultCode == RESULT_OK) {
                notifyDataChanged()
            }
        }
    }

    interface OnBeansTypeSelectedListener {
        fun onBeansTypeSelected(beansType: BeansType)
    }

    override fun getDataList(): List<BeansType> {
        return beansTypeDataManager.getBeansTypes()
    }

    override fun createContentViewHolder(): ContentViewHolder<BeansType> {
        return object : ContentViewHolder<BeansType>() {
            private val beansNameTextView: TextView
            private val beansCountryTextView: TextView

            init {
                val inflater = LayoutInflater.from(context)
                view = inflater.inflate(R.layout.item_beans_type_content_view, null, false)

                beansNameTextView =
                        view.findViewById(R.id.item_beans_type_row_name_text_view) as TextView
                beansCountryTextView =
                        view.findViewById(R.id.item_beans_type_row_country_text_view) as TextView
            }

            override fun onClicked() {
                if (!isInEditMode) {
                    onBeansTypeSelectedListener.onBeansTypeSelected(data)
                } else {
                    val intent = BeansTypeActivity.newIntent(context, data.id)
                    startActivityForResult(intent, EDIT_BEANS_TYPE_REQUEST)
                }
            }

            override fun onUpdateView() {
                beansNameTextView.text = data.name
                beansCountryTextView.text = data.country
                if (data.id == highlightedBeansTypeId) {
                    beansNameTextView.setTypeface(null, Typeface.BOLD)
                }
            }

            override fun onRecycle() {
                if (data.id == highlightedBeansTypeId) {
                    beansNameTextView.setTypeface(null, Typeface.NORMAL)
                }
            }
        }
    }

    override fun onRemoveContentViewHolder(contentViewHolder: ContentViewHolder<BeansType>) {
        // TODO: remove this tight coupling to the data manager. Create a recyclerView,
        // which operates on Cursor instead.
        beansTypeDataManager.removeBeansType(contentViewHolder.data)
        adapter.setDataList(beansTypeDataManager.getBeansTypes())
    }

    companion object {
        private val ARG_HIGHLIGHTED_BEANS_TYPE_ID = "highlighted_beans_type_id"
        private val EDIT_BEANS_TYPE_REQUEST = 1

        fun newInstance(highlightedBeansTypeId: UUID): BeansTypeListFragment {
            val args = Bundle()
            args.putSerializable(ARG_HIGHLIGHTED_BEANS_TYPE_ID, highlightedBeansTypeId)
            val beansTypeListFragment = BeansTypeListFragment()
            beansTypeListFragment.arguments = args
            return beansTypeListFragment
        }

        fun newInstance(): BeansTypeListFragment {
            return BeansTypeListFragment()
        }
    }

}
